### PR TITLE
Unknown component

### PR DESCRIPTION
**Pull request by So-Fras**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#506

### CODE REVIEW

File: single-line-diagram/single-line-diagram-core/src/main/resources/ConvergenceLibrary/unknown-component.svg [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg xmlns="http://www.w3.org/2000/svg" version="1.0" height="16" width="16">
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/main/resources/ConvergenceLibrary/unknown-component.svg:2]: Fix width and height to 20, even if they are ignored when generating the diagram it's used when previewing the component

File: single-line-diagram/single-line-diagram-core/src/main/resources/ConvergenceLibrary/unknown-component.svg [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+    <path d="M 9.9248465,0 0,10.015385 9.9556158,20.030769 19.999391,9.9538462 Z m 2.5705015,7.9302308 c 0,1.7033846 -0.851881,1.9296154 -2.196256,2.8611542 v 1.211 H 9.2741726 v -1.67677 C 10.432199,9.7267693 11.49705,9.261 11.470429,7.9967693 11.443808,6.3466154 8.8866164,6.577 7.6221055,7.3621539 L 7.9297978,7.3313847 V 6.1869231 C 9.9663259,5.5481539 12.495348,5.588077 12.495348,7.9302308 Z M 9.2434034,12.671154 h 1.1047836 v 1.224308 H 9.2434034 Z"/>
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/main/resources/ConvergenceLibrary/unknown-component.svg:3]: Precision is too high for what is needed, I simplified it with a precision of 3:
```suggestion
    <path d="M 10,0 -0,10 10,20 20,10 Z m 2.57,7.93 c 0,1.7 -0.8,1.93 -2.2,2.87 V 12 H 9.28 V 10.3 C 10.5,9.73 11.5,9.26 11.5,8 11.5,6.35 8.89,6.58 7.62,7.36 H 7.93 V 6.19 C 9.97,5.55 12.5,5.59 12.5,7.93 Z M 9.25,12.7 h 1.15 v 1.2 H 9.25 Z"/>
```

File: single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        Mockito.when(componentLibrary.getSubComponentStyleClass(ComponentTypeName.THREE_WINDINGS_TRANSFORMER, "WINDING3")).thenReturn(unknownSubComponentStyleClass);
+
+        // write SVGs and compare to reference
+        assertEquals(toString("/TestUnknownConvergence.svg"), toSVG(g1, "/TestUnknownConvergence.svg"));
+    }
+
+    private void buildNetwork() {
+        // first voltage level
+        //
+        vl1 = createVoltageLevel(substation, "vl1", "vl1", TopologyKind.NODE_BREAKER, 380, 50);
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java:88]: Duplicate code with other unit test, you can put this in NetworkFactory or CreateNetworksUtil (I know, this two factory files should be merged or renamed...)

File: single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+        Mockito.when(componentLibrary.getSvgElements(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(unknownSvgElements);
+        Mockito.when(componentLibrary.getComponentStyleClass(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(unknownComponentStyleClass);
+        Mockito.when(componentLibrary.getSize(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(unknownSize);
+        Mockito.when(componentLibrary.getTransformations(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(unknownTransformations);
+        Mockito.when(componentLibrary.getAnchorPoints(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(unknownAnchorPoints);
+        Mockito.when(componentLibrary.getSubComponentStyleClass(ComponentTypeName.THREE_WINDINGS_TRANSFORMER, "WINDING1")).thenReturn(unknownSubComponentStyleClass);
+        Mockito.when(componentLibrary.getSubComponentStyleClass(ComponentTypeName.THREE_WINDINGS_TRANSFORMER, "WINDING2")).thenReturn(unknownSubComponentStyleClass);
+        Mockito.when(componentLibrary.getSubComponentStyleClass(ComponentTypeName.THREE_WINDINGS_TRANSFORMER, "WINDING3")).thenReturn(unknownSubComponentStyleClass);
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java:79]: You should use mockito to simulate one component missing, so you should only need:
```java
Mockito.when(componentLibrary.getSvgElements(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)).thenReturn(null);
```

File: single-line-diagram/single-line-diagram-core/src/main/resources/FlatDesignLibrary/unknown-component.svg [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<svg xmlns="http://www.w3.org/2000/svg" version="1.0" height="16" width="16">
+    <path d="M 9.8976923,0 0.06153846,9.99 9.9584615,20 20,10.02 Z m 1.2061537,11.034615 v 1.450001 H 8.8838461 v -2.160001 c 3.9699999,-0.8799997 2.6299999,-5.0199997 -1.24,-2.5599997 h -0.26 v -1.99 c 5.8699999,-2.31 8.8699999,2.91 3.7199999,5.2599997 z m -2.3907691,2.411539 h 2.5300001 v 1.8 H 8.7130769 Z"/>
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/main/resources/FlatDesignLibrary/unknown-component.svg:3]: Same comments 
```suggestion
    <path d="M 10,0 0,10 10,20 20,10 Z m 1.2,11 v 1.5 H 8.89 V 10.3 C 12.9,9.44 11.5,5.3 7.65,7.76 H 7.39 V 5.77 C 13.3,3.46 16.3,8.68 11.1,11 Z m -2.38,2.4 h 2.58 v 1.8 H 8.72 Z" />
```

**Review by flo-dup — CHANGES_REQUESTED** [review]

File: single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java (hunk 4 above)
Comment (tadam50) [reply single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java:79]: In ResourcesComponentLibrary :
We also need to simulate returning `null` from calls of : `components.get(ComponentTypeName.THREE_WINDINGS_TRANSFORMER)`
Or returning 'unknown comonent' as following : components.get(ComponentTypeName.UNKNOWN_COMPONENT) ?

File: single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/library/ResourcesComponentLibrary.java [hunk 6]
```diff
@@ -170,8 +174,25 @@ public Optional<String> getSubComponentStyleClass(String type, String subCompone
...
+    protected Component getComponent(String type) {
+        Objects.requireNonNull(type);
+        Component component = components.get(type);
+        if (component == null && isInsideVlComponent(type)) {
+            component = components.get(ComponentTypeName.UNKNOWN_COMPONENT);
+        }
+        return component;
+    }
+
+    protected boolean isInsideVlComponent(String componentType) {
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/library/ResourcesComponentLibrary.java:190]: Could you rename this to `isDisplayed`?
I think we should address this in another way later : adding the "invisible" components in the `components.json`.

**Review by flo-dup — CHANGES_REQUESTED** [review]

File: single-line-diagram/single-line-diagram-core/src/test/resources/UnknownLibrary/battery.svg
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/test/resources/UnknownLibrary/battery.svg:1]: You can remove this component from this library, in fact all components except load, breaker, disconnector and arrows.

File: single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (flo-dup) [inline single-line-diagram/single-line-diagram-core/src/test/java/com/powsybl/sld/iidm/TestUnknownComponent.java:2]: Please fix the copyright year and add  `* SPDX-License-Identifier: MPL-2.0`

**Review by flo-dup — CHANGES_REQUESTED** [review]
Still two minor comments and we're good!

**Review by flo-dup — APPROVED** [review]

**Pull request merged**